src/coreclr/tools/superpmi/superpmi-shared/methodcontext.cpp

```diff
@@ -6683,117 +6443,174 @@ CORINFO_CLASS_HANDLE MethodContext::repGetTypeInstantiationArgument(CORINFO_CLAS
     return result;
 }
 
-void MethodContext::recAppendClassName(
-    int nBufLenIn,
-    CORINFO_CLASS_HANDLE cls,
-    bool fNamespace,
-    bool fFullInst,
-    bool fAssembly,
-    int nLenOut,
-    const char16_t* result)
-{
-    if (AppendClassName == nullptr)
-        AppendClassName = new LightWeightMap<Agnostic_AppendClassNameIn, Agnostic_AppendClassNameOut>();
-
-    // The API has two different behaviors depending on whether the input specified length is zero or non-zero:
-    // (1) zero: returns the length of the string (which the caller can use to allocate a buffer)
-    // (2) non-zero: fill as much of the buffer with the name as there is space available.
-    // We don't want the input length to be part of the key, since the caller could potentially pass in a smaller
-    // or larger buffer, and we'll fill in as much of the buffer as possible. We do need to handle both the zero
-    // and non-zero cases, though, so we'll get one record for first call using zero (with the correct buffer size result),
-    // and another for the second call with a big-enough buffer. We could presumably just store the second case
-    // (and overwrite the first record), since it contains the same output length, but it is useful to store
-    // (and see) all the JIT-EE interface calls.
-
-    Agnostic_AppendClassNameIn key;
-    ZeroMemory(&key, sizeof(key)); // Zero key including any struct padding
-    key.nBufLenIsZero = (nBufLenIn == 0) ? 1 : 0;
-    key.classHandle   = CastHandle(cls);
-    key.fNamespace    = fNamespace;
-    key.fFullInst     = fFullInst;
-    key.fAssembly     = fAssembly;
+void MethodContext::recPrint(
+    const char* name,
+    LightWeightMap<DWORDLONG, Agnostic_PrintResult>*& map,
+    DWORDLONG handle,
+    char* buffer,
+    size_t bufferSize,
+    size_t* pRequiredBufferSize,
+    size_t bytesWritten)
+{
+    if (map == nullptr)
+        map = new LightWeightMap<DWORDLONG, Agnostic_PrintResult>();
+
+    // Required size of a buffer that contains all data and null terminator.
+    UINT requiredBufferSize = UINT_MAX;
+    if (pRequiredBufferSize != nullptr)
+    {
+        requiredBufferSize = (UINT)(*pRequiredBufferSize);
+    }
+    else if (bytesWritten + 1 < bufferSize)
+    {
+        requiredBufferSize = (UINT)(bytesWritten + 1);
+    }
+
+    Agnostic_PrintResult res;
+    int index = map->GetIndex(handle);
+    if (index != -1)
+    {
+        // Merge with existing entry
+
+        res = map->GetItem(index);
+
+        if (requiredBufferSize != UINT_MAX)
+        {
+            res.requiredBufferSize = requiredBufferSize;
+        }
+
+        if (bytesWritten > res.bufferSize)
+        {
+            // Always stored without null terminator.
+            res.buffer = map->AddBuffer((unsigned char*)buffer, static_cast<unsigned>(bytesWritten));
+            res.bufferSize = (UINT)bytesWritten;
```

Comment:
```suggestion
            res.bufferSize = (UINT)(bytesWritten + 1);
```
?

Reply:
There's something confusing here where you aren't actually storing the argument `bufferSize`, which makes sense, but what you are storing as `bufferSize` doesn't have quite the same semantics. Maybe the stored `bufferSize` should get a different name, such as `bufferBytesWritten`?

Reply:
I can understand the confusion. The `Agnostic_PrintResult::bufferSize` is actually just the size of the buffer that `Agnostic_PrintResult::buffer` is pointing at. So it is not really related to the buffer size passed to these functions.
The string is always stored without null terminator.

With the unified string handling, I've spent the extra effort to implement this in a way where we don't need to split the key on the buffer size, and simply always store the "maximal" information we queried during recording. The benefit is that we are able to replay for any `bufferSize` below that (or above it, if we have all data) without missing.

I think to make this less confusing I will rename to `Agnostic_PrintResult::stringBuffer` and `Agnostic_PrintResult::stringBufferSize` to avoid the conflict with the `bufferSize` parameter in these function. With that in mind, does the string handling make more sense in regards to the null terminators?

Reply:
> I think to make this less confusing I will rename to `Agnostic_PrintResult::stringBuffer` and `Agnostic_PrintResult::stringBufferSize` to avoid the conflict with the `bufferSize` parameter in these function. With that in mind, does the string handling make more sense in regards to the null terminators?

Yes, as long as there are good comments that specify if the `stringBufferSize` includes the null terminator or not.